File: src/engine/search/discovery.py
"""
KIVO Search Engine
Discovery

Die groesste Architekturentscheidung aus der Spec: die Engine besitzt
KEIN eingebautes Weltwissen. Sie beobachtet, welche Begriffe in mehreren
Entries auftauchen, und erklaert die selbst zu "Konzepten".

Kein Docker/Caddy/Kubernetes-Woerterbuch - das entsteht rein aus den
Daten des Nutzers.
"""

from __future__ import annotations

from .entry import Entry
from .analysis import Analysis, ConceptStats
from .pipeline.tokenizer import Tokenizer
from .pipeline.normalizer import Normalizer
from .text_cleaning import strip_media


def discover(entries: tuple[Entry, ...], stopwords: set[str] | None = None, normalizer=None) -> Analysis:
    """Baut eine komplett neue Analysis aus dem aktuellen Entry-Bestand."""

    tokenizer = Tokenizer()
    active_normalizer = normalizer if normalizer is not None else Normalizer()
    active_stopwords = stopwords if stopwords is not None else _stopwords()

    analysis = Analysis()
    analysis._total_documents = len(entries)

    for entry in entries:
        text = strip_media(f"{entry.title} {entry.content}")
        raw_tokens = tokenizer.tokenize(text)
        filtered = [t for t in raw_tokens if t not in active_stopwords]
        normalized = [active_normalizer.normalize(t) for t in filtered]

        seen_in_this_entry: set[str] = set()

        for term in normalized:
            if len(term) < 2:
                continue

            stats = analysis.concepts.setdefault(term, ConceptStats(term=term))
            stats.total_frequency += 1

            if term not in seen_in_this_entry:
                stats.document_frequency += 1
                stats.entry_ids.add(entry.id)
                seen_in_this_entry.add(term)

    return analysis


def _stopwords():
    from .pipeline.stopwords import STOPWORDS
    return STOPWORDS

File: src-tauri/src/engine/discovery.rs
//! KIVO Search Engine (Rust-Port)
//! Discovery
//!
//! 1:1-Port von `src/engine/search/discovery.py`.
//!
//! Die groesste Architekturentscheidung aus der Python-Spec bleibt
//! erhalten: die Engine besitzt KEIN eingebautes Weltwissen. Sie
//! beobachtet, welche Begriffe in mehreren Entries auftauchen, und
//! erklaert die selbst zu "Konzepten".
//!
//! SPRACH-HINWEIS: `stopwords` ist jetzt parametrisiert (siehe
//! `pipeline::stopwords::stopwords_for`) - die Sprachumschaltung "auto"/
//! "de"/"en" wirkt sich also auf die Stopword-Filterung aus. Der
//! Normalizer bleibt weiterhin immer der heuristische (kein Snowball-
//! Stemmer-Support) - siehe Scope-Hinweis in `pipeline::stopwords`.

use std::collections::HashSet;

use crate::engine::analysis::{Analysis, ConceptStats};
use crate::engine::entry::Entry;
use crate::engine::pipeline::normalizer;
use crate::engine::pipeline::tokenizer;
use crate::engine::text_cleaning::strip_media;

/// Entspricht `discover(entries, stopwords=...)`. `stopwords` steuert die
/// Sprachumschaltung (siehe `pipeline::stopwords::stopwords_for`).
pub fn discover(entries: &[Entry], stopwords: &HashSet<&'static str>) -> Analysis {
    let mut analysis = Analysis::new();
    analysis.total_documents = entries.len();

    for entry in entries {
        let text = strip_media(&format!("{} {}", entry.title, entry.content));
        let raw_tokens = tokenizer::tokenize(&text);
        let filtered: Vec<String> = raw_tokens
            .into_iter()
            .filter(|t| !stopwords.contains(t.as_str()))
            .collect();
        let normalized: Vec<String> = filtered.iter().map(|t| normalizer::normalize(t)).collect();

        let mut seen_in_this_entry: HashSet<String> = HashSet::new();

        for term in normalized {
            if term.chars().count() < 2 {
                continue;
            }

            let stats = analysis
                .concepts
                .entry(term.clone())
                .or_insert_with(|| ConceptStats::new(term.clone()));
            stats.total_frequency += 1;

            if !seen_in_this_entry.contains(&term) {
                stats.document_frequency += 1;
                stats.entry_ids.insert(entry.id);
                seen_in_this_entry.insert(term);
            }
        }
    }

    analysis
}
